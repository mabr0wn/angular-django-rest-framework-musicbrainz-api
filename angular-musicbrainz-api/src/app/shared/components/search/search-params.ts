export class SearchParams {
    term: string;
    type: string;
    constructor(term, type) {
      this.term = term;
      this.type = type;
    }

    equals(params: SearchParams): boolean {
      return this.term === params.term && this.type === params.type;
    }
  }
