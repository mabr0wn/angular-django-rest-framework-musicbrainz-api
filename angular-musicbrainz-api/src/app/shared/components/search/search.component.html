<input 
  matInput 
  id="search-input" 
  [matAutocomplete]="auto" 
  [formControl]="searchControl" 
  [placeholder]="searchType === 'release' ? 'Album Title' : 'Artist Name'" 
  (keyup)="searchFor(search.value)" 
  name="search" 
  minlength="3" 
  required 
  #search>
<mat-autocomplete  #auto="matAutocomplete">
  <mat-option *ngFor="let item of filteredResults$ | async" [value]="item">
      <span [innerHTML]="item | bold:searchControl.value"></span>
  </mat-option>
</mat-autocomplete>
