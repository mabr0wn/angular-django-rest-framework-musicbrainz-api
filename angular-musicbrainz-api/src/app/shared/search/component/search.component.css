@import '@angular/material/prebuilt-themes/deeppurple-amber.css';


:host {
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    height: 100%;
  } 
  
  h1 {
    font-weight: 300;
    font-size: 40px;
    color: rgba(0, 0, 0, 0.87);
  }
  
  input {
    border: solid 1px rgba(0, 0, 0, 0.1);
    border-radius: 2px;
    outline: none;
    padding: 6px 16px;
  
    width: 100%;
    max-width: 375px;
  
    font-size: 16px;
    line-height: 34px;
  }
  .fa-search {
    margin-left: 27rem;
    margin-top: -2.25rem;
  }
  
  .spacer {
    height: 200px;
  }

  footer {
    max-width: 100vw;
  }
  
  .container {
    display: flex;
    flex-direction: column;
  }
  
  .row {
    width: 60%;
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    padding: 1em;
    align-items: flex-start;
    margin: auto;
  }
  
  .row-item {
    display: flex;
    width: calc(33% - 0.5em);
    padding: 0.5em 2em;
  }
  
  .row-item img {
    max-width: 50%;
    height: auto;
    display: block;
    margin: auto;
  }
  
  .row-item p {
    font-size: 0.8em;
  }
  
  .row-item.one {
    justify-content: flex-end;
  }
  
  .row-item.one p {
    text-align: end;
    text-align: right;
    -webkit-text-align: right;
  }
  
  .row-item.two {
    justify-content: center;
    border-right: 1px solid #ccc;
    border-left: 1px solid #ccc;
  }
  
  .row-item.three {
    justify-content: flex-start;
  }
  
  /* Something Fun */
  #countMe:hover {
    animation-name: wiggle;
    -webkit-animation-name: wiggle;
    -ms-animation-name: wiggle;
    -ms-animation-duration: 1000ms;
    -webkit-animation-duration: 1000ms;
    -webkit-animation-iteration-count: 1;
    -ms-animation-iteration-count: 1;
    -webkit-animation-timing-function: ease-in-out;
    -ms-animation-timing-function: ease-in-out;
  }
  @keyframes wiggle {
    0% {transform: rotate(10deg);}
    25% {transform: rotate(-10deg);}
    50% {transform: rotate(20deg);}
    75% {transform: rotate(-5deg);}
    100% {transform: rotate(0deg);}
  }
  
  
  