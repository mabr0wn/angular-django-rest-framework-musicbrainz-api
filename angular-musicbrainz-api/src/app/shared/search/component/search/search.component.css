@import '@angular/material/prebuilt-themes/deeppurple-amber.css';


:host {
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    height: 100%;
  } 
  
  h1 {
    font-weight: 300;
    font-size: 40px;
    color: rgba(0, 0, 0, 0.87);
  }
  
  input {
    border: solid 1px rgba(0, 0, 0, 0.1);
    border-radius: 2px;
    outline: none;
    padding: 6px 16px;
  
    width: 100%;
    max-width: 375px;
  
    font-size: 16px;
    line-height: 34px;
  }
  .fa-search {
    margin-left: 27rem;
    margin-top: -2.25rem;
  }
  
  .spacer {
    height: 200px;
  }
  
  