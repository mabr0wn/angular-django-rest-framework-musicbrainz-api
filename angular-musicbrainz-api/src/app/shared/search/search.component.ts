// Angular
import { Component, OnInit } from '@angular/core';
import { FormControl } from '@angular/forms';
// RxJs
import { Observable } from 'rxjs';
import { map, startWith } from 'rxjs/operators';
// Dummy data
import { SAMPLE_RESULTS } from '../../sample-results';


@Component({
  selector: 'app-search',
  templateUrl: './search.component.html',
  styleUrls: ['./search.component.css']
})
export class SearchComponent implements OnInit {
  searchControl: FormControl;

  filteredResults$: Observable<string[]>;

  results = SAMPLE_RESULTS;

  constructor() { }

  ngOnInit() {
    this.searchControl = new FormControl('');
    this.filteredResults$ = this.searchControl.valueChanges.pipe(
      startWith(''),
      map(val => this.filterResults(val)),
      map(val => val.slice(0, 4)));
  }

  filterResults(val: string): string[] {
    return val ? this.results.filter(v => v.toLowerCase().indexOf(val.toLowerCase()) === 0) : [];
  }

}
