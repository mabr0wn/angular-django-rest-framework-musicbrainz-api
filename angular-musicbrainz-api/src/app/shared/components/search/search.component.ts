// Angular
import {
  Component,
  OnInit,
  OnDestroy,
  Output,
  EventEmitter,
  ChangeDetectionStrategy
 } from '@angular/core';
import { FormControl } from '@angular/forms';
// RxJs
import { 
  Observable, 
  Subject
} from 'rxjs';
import {
  map,
  startWith, 
  takeUntil 
} from 'rxjs/operators';
// Dummy data
import { SAMPLE_RESULTS } from '../../../sample-results';
// Model
import { Album } from '@core/models/album';
// Local
import { SearchParams} from '@shared/search-params';
import { SearchPresenter } from '@shared/presenters/search.presenter';

@Component({
  selector: 'app-search-component',
  templateUrl: './search.component.html',
  styleUrls: ['./search.component.css'],
  changeDetection: ChangeDetectionStrategy.OnPush,
  providers: [SearchPresenter],
})
export class SearchComponent implements OnInit, OnDestroy {
  @Output() search: EventEmitter<SearchParams> = new EventEmitter<SearchParams>();
  private ngUnsubscribe: Subject<any> = new Subject();
  //
  queryString: string;
  searching: boolean;
  searchType: string;
  searchTerms: Subject<SearchParams> = new Subject<SearchParams>();
  //
  searchControl: FormControl;
  filteredResults$: Observable<string[]>;
  results = SAMPLE_RESULTS;


  constructor(private presenter: SearchPresenter) { }

  // ngOnInit(): void {
  //   this.searchControl = new FormControl('');
  //   this.filteredResults$ = this.searchControl.valueChanges.pipe(
  //     startWith(''),
  //     map(val => this.filterResults(val)),
  //     map(val => val.slice(0, 4)));
  // }

  ngOnInit(): void {
    this.presenter.searchTerms$.pipe(
      // complete when component is destroyed
      takeUntil(this.searchTerms),
    ).subscribe(() => this.search.emit());
  }

  searchFor(): void {
    this.queryString = this.queryString ? this.queryString.trim() : null;
    if (this.queryString) {
      this.searching = true;
      this.presenter.search();
      this.queryString = null;
    }
  }

  filterResults(val: string): string[] {
    return val ? this.results.filter(v => v.toLowerCase().indexOf(val.toLowerCase()) === 0) : [];
  }

  ngOnDestroy() {
    this.ngUnsubscribe.next();
    this.ngUnsubscribe.complete();
    this.searchTerms.next();
    this.searchTerms.complete();
  }

}
