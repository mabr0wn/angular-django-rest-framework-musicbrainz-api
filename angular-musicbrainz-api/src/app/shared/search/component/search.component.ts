// Angular
import {
  Component,
  OnInit,
  OnDestroy,
  Output,
  EventEmitter,
  ChangeDetectionStrategy,
  Input
 } from '@angular/core';
import { FormControl } from '@angular/forms';
// RxJs
import { 
  Observable, 
  Subject
} from 'rxjs';
import { 
  map,
  startWith,
  takeUntil 
} from 'rxjs/operators';
// Dummy data
import { SAMPLE_RESULTS } from '../../../sample-results';
// Model
import { Album } from '@core/models/album';
// Local
import { SearchParams} from '@shared/search/search-params';
import { SearchPresenter } from '@shared/search/presenter/search.presenter';

@Component({
  selector: 'app-search-component',
  templateUrl: './search.component.html',
  styleUrls: ['./search.component.css'],
  changeDetection: ChangeDetectionStrategy.OnPush,
  providers: [SearchPresenter],
})
export class SearchComponent implements OnInit, OnDestroy {
  @Input() albums: Album[];
  @Input() queryString: string;
  @Output() search: EventEmitter<SearchParams> = new EventEmitter();

  private ngUnsubscribe: Subject<any> = new Subject();
  filteredResults$: Observable<string[]>;
  searchControl: FormControl;

  results = SAMPLE_RESULTS;
  searchType: string;

  constructor(private presenter: SearchPresenter) {}

  ngOnInit(): void {
    this.searchControl = new FormControl('');
    this.filteredResults$ = this.searchControl.valueChanges.pipe(
      startWith(''),
      map(val => this.filterResults(val)),
      map(val => val.slice(0, 4)));
    this.presenter.searchTerms.pipe(
      // complete when component is destroyed
      takeUntil(this.ngUnsubscribe),
    ).subscribe(_ => this.search.emit(new SearchParams(this.queryString, this.searchType)));
  }

  searchFor(term): void {
    // Log the term for debugging purposes
    console.log(term, '@component');
    this.presenter.search(term);
  }

  filterResults(val: string): string[] {
    return val ? this.results.filter(v => v.toLowerCase().indexOf(val.toLowerCase()) === 0) : [];
  }

  ngOnDestroy(): void {
    this.ngUnsubscribe.next();
    this.ngUnsubscribe.complete();
  }
}
