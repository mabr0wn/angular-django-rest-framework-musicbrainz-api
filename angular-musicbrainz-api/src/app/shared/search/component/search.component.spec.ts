// Angular Testing
import {
  async,
  ComponentFixture,
  TestBed
} from '@angular/core/testing';
// RxJs
import { of } from 'rxjs';
// Components
import { SearchComponent } from './search.component';
// Pipe
import { BoldPipe } from '@core/pipe/bold.pipe';
// Services
import { SearchService } from '@core/services/search/search.service';
// Models
import { Album } from '@core/models/album';
// Modules
import {
  MatAutocompleteModule,
  MatFormFieldModule
} from '@angular/material';
import {
  FormsModule,
  ReactiveFormsModule,
} from '@angular/forms';
import { MDBBootstrapModule } from 'angular-bootstrap-md';


describe('SearchComponent', () => {
  let component: SearchComponent;
  let fixture: ComponentFixture<SearchComponent>;
  let dummyResult: Album[];
  const searchService = {
    queryAlbums: jest.fn()
  };

  beforeEach(async(() => {

    dummyResult = [
      {id: '1', title: 'title1', artistCredit: [], image: 'image1'},
      {id: '2', title: 'title2', artistCredit: [], image: 'image2'},
    ];

    TestBed.configureTestingModule({
      imports: [
        MDBBootstrapModule,
        MatAutocompleteModule,
        MatFormFieldModule,
        FormsModule,
        ReactiveFormsModule
      ],
      declarations: [
        BoldPipe,
        SearchComponent
      ],
      providers: [
        {provide: SearchService, useValue: searchService},
      ]
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(SearchComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  describe('Should run "searchFor" method', () => {
    const mockQueryString = 'testQuery';
    let setUp = () => {
      searchService.queryAlbums.mockReturnValueOnce(of(dummyResult));
      component.query = mockQueryString;
    }
    test('calls searchService.queryAlbums with queryString', () => {
      setUp();
      expect(searchService.queryAlbums).not.toHaveBeenCalled();

      component.searchFor();
      expect(searchService.queryAlbums).toMatchSnapshot(mockQueryString, component.searchType);
    });
  });
});
