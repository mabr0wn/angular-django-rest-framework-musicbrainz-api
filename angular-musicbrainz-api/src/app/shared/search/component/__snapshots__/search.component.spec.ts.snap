// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchComponent Should run "searchFor" method calls searchService.queryAlbums with queryString: testQuery 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "testQuery",
      "release",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Observable {
        "_isScalar": false,
        "_subscribe": [Function],
      },
    },
  ],
}
`;
